lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,468 @@
+                (KTaskNum % config.KThreads) * 1.0 / KTaskNum;
+  if (MTaskNum < config.MThreads || NTaskNum < config.NThreads ||
+      KTaskNum < config.KThreads) {
+    double threadNotFulllyUtilizedPenalty = 10.0;
+    cost *= threadNotFulllyUtilizedPenalty;
+  }
+  return cost;
+}
+
+// calculate the cost of the memory consumption on the thread
+double memoryConsumptionOnThreadCost(linalg::LinalgOp &linalgOp,
+                                     ArrayRef<uint32_t> shape,
+                                     const MatmulConfig &config,
+                                     SystemDesc &sysDesc) {
+  if (shape.size() < 3) {
+    // Has an invalid shape
+    return 0;
+  }
+  uint32_t M = shape[0], N = shape[1], K = shape[2];
+  size_t dtypeSize = DataLayout().getTypeSize(
+      ShapeAdaptor(linalgOp.getDpsInputs()[1].getType()).getElementType());
+  // if use K split, there will be one more final reduce and break the post
+  // fusion
+  double KSplitPenalty = 8.0 * dtypeSize;
+  double memoryConsumptionPerThread =
+      M * K * 1.0 / config.MThreads / config.KThreads +
+      K * N * 1.0 / config.KThreads / config.NThreads +
+      M * N * ((config.KThreads - 1) * KSplitPenalty + 1.0) / config.MThreads /
+          config.NThreads;
+  return memoryConsumptionPerThread;
+}
+
+// calculate the cost of the computation intensity on the L2 cache
+double computationIntensityOnL2Cache(linalg::LinalgOp &linalgOp,
+                                     ArrayRef<uint32_t> shape,
+                                     const MatmulConfig &config,
+                                     SystemDesc &sysDesc) {
+  double fullLoadRatio = 0.7;
+  uint32_t L2Cache = sysDesc.getCacheSize(2);
+  size_t dtypeSize = DataLayout().getTypeSize(
+      ShapeAdaptor(linalgOp.getDpsInputs()[1].getType()).getElementType());
+  uint32_t outOfCachePenalty = 1024;
+  double FLOPS = 2.0 * config.MBlock * config.NBlock * config.KBlock;
+  double memoryConsumption = config.MBlock * config.NBlock +
+                             config.NBlock * config.KBlock +
+                             config.MBlock * config.KBlock;
+  double computationIntensity = FLOPS / memoryConsumption;
+  if (memoryConsumption * dtypeSize > L2Cache * fullLoadRatio) {
+    computationIntensity /= outOfCachePenalty;
+  }
+  return 1 / computationIntensity;
+}
+
+using CostModelFn =
+    std::function<double(linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                         MatmulConfig cfg, SystemDesc &sysDesc)>;
+
+// filter the config by the cost model
+std::vector<MatmulConfig>
+filterConfigByCostModel(ArrayRef<MatmulConfig> configs,
+                        linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                        SystemDesc &sysDesc, const CostModelFn &costModel,
+                        float eliminationRatio = 0.5, float threshold = -1) {
+  std::vector<MatmulConfig> result;
+  std::vector<float> costs;
```

Comment:
What is the meaning of `eliminationRatio` here?

Reply:
Change it `preserveRatio` to reduce the confusion. It means that it only preserves the top `preserveRatio`(30% for example) candidates and the rest will be thrown away.